### PR TITLE
Cladtorch and GPT-2 implementation

### PR DESCRIPTION
**Pull request by r-timmaraju**
Adds implementation for the tensor header-only library `cladtorch`, and an implementation of the GPT-2 architecture using it. The training loop fine-tunes on the tiny-shakespeare dataset and uses `clad::gradient` for the backpropagation pass.

### CODE REVIEW

**Comment by r-timmaraju** [comment]
@vgvassilev I worked around the issue that was happening before by adding custom derivatives for everything - it should now compile properly. Only issue that's still occurring is for some reason the CI can't link to openmp (maybe that's an issue with the CMake config I made but not fully sure).

Other than that I think it's ready to merge.

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 1]
```diff
@@ -0,0 +1,1299 @@
...
+                          ::cladtorch::Tensor<T>* _d_other) {
+  *_d_other += *_d_this;
+  _d_this->fill(0);
+}
+
+template <typename T>
+::clad::ValueAndPushforward<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
+constructor_pushforward(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
+                        ::cladtorch::Tensor<T>&& p,
+                        ::cladtorch::Tensor<T>&& d_p) {
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h:1073]:
```suggestion
                        ::cladtorch::Tensor<T>& d_p) {
```

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 2]
```diff
@@ -0,0 +1,1299 @@
...
+  ::cladtorch::Tensor<T> v(::std::move(p));
+  ::cladtorch::Tensor<T> d_v(::std::move(d_p));
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h]:
```suggestion
  ::cladtorch::Tensor<T> d_v(d_p);
```

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 3]
```diff
@@ -0,0 +1,1299 @@
...
+  ::cladtorch::Tensor<T> v(p);
+  ::cladtorch::Tensor<T> d_v(d_p);
+  return {v, d_v};
+}
+
+template <typename T>
+::clad::ValueAndAdjoint<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
+constructor_reverse_forw(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
+                         ::cladtorch::Tensor<T>&& p,
+                         ::cladtorch::Tensor<T>&& d_p) {
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h:1118]:
```suggestion
                         ::cladtorch::Tensor<T>& d_p) {
```

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 4]
```diff
@@ -0,0 +1,1299 @@
...
+  ::cladtorch::Tensor<T> v(::std::move(p));
+  ::cladtorch::Tensor<T> d_v(::std::move(d_p));
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h]:
```suggestion
  ::cladtorch::Tensor<T> d_v(d_p);
```

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 5]
```diff
@@ -0,0 +1,1299 @@
...
+
+template <typename T>
+::clad::ValueAndPushforward<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
+constructor_pushforward(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
+                        ::cladtorch::Tensor<T>&& p,
+                        ::cladtorch::Tensor<T>& d_p) {
+  ::cladtorch::Tensor<T> v(::std::move(p));
+  ::cladtorch::Tensor<T> d_v(d_p);
+  return {v, d_v};
+}
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h:1077]:
```suggestion
template <typename T>
::clad::ValueAndPushforward<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
constructor_pushforward(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
                        ::cladtorch::Tensor<T>&& p,
                        ::cladtorch::Tensor<T>&& d_p) {
  ::cladtorch::Tensor<T> v(::std::move(p));
  ::cladtorch::Tensor<T> d_v(d_p);
  return {v, d_v};
}
```

File: include/clad/Differentiator/CladtorchBuiltins.h [hunk 6]
```diff
@@ -0,0 +1,1299 @@
...
+
+template <typename T>
+::clad::ValueAndAdjoint<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
+constructor_reverse_forw(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
+                         ::cladtorch::Tensor<T>&& p,
+                         ::cladtorch::Tensor<T>& d_p) {
+  ::cladtorch::Tensor<T> v(::std::move(p));
+  ::cladtorch::Tensor<T> d_v(::std::move(d_p));
+  return {v, d_v};
+}
```
Comment (fogsong233) [inline include/clad/Differentiator/CladtorchBuiltins.h:1122]:
```suggestion
template <typename T>
::clad::ValueAndAdjoint<::cladtorch::Tensor<T>, ::cladtorch::Tensor<T>>
constructor_reverse_forw(ConstructorPushforwardTag<::cladtorch::Tensor<T>>,
                         ::cladtorch::Tensor<T>&& p,
                         ::cladtorch::Tensor<T>&& d_p) {
  ::cladtorch::Tensor<T> v(::std::move(p));
  ::cladtorch::Tensor<T> d_v(d_p);
  return {v, d_v};
}
```
